File: backend/src/entities/Participant.ts
import {
  Entity, PrimaryGeneratedColumn, Column, CreateDateColumn, UpdateDateColumn,
  ManyToOne, OneToMany, OneToOne, JoinColumn, Unique,
} from 'typeorm';
import { Event } from './Event.js';
import { NfcTag } from './NfcTag.js';
import { PointTransaction } from './PointTransaction.js';
import { Redemption } from './Redemption.js';

@Entity('participants')
@Unique(['eventId', 'displayName'])
export class Participant {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ length: 100 })
  displayName: string;

  @Column({ length: 100, nullable: true })
  firstName: string | null;

  @Column({ length: 100, nullable: true })
  lastName: string | null;

  @Column({ type: 'int', nullable: true })
  age: number | null;

  @Column({ length: 100, nullable: true })
  group: string | null;

  @Column({ type: 'jsonb', default: {} })
  emergencyData: EmergencyData;

  @Column({ type: 'jsonb', default: {} })
  customFields: Record<string, unknown>;

  @Column({ type: 'int', default: 0 })
  balance: number;

  @Column({ default: true })
  active: boolean;

  @ManyToOne(() => Event, (e) => e.participants, { onDelete: 'CASCADE' })
  @JoinColumn({ name: 'eventId' })
  event: Event;

  @Column('uuid')
  eventId: string;

  @OneToOne(() => NfcTag, (t) => t.participant, { nullable: true })
  nfcTag: NfcTag | null;

  @OneToMany(() => PointTransaction, (t) => t.participant)
  transactions: PointTransaction[];

  @OneToMany(() => Redemption, (r) => r.participant)
  redemptions: Redemption[];

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;
}

export interface EmergencyData {
  bloodType?: string;
  allergies?: string[];
  medications?: string[];
  conditions?: string[];
  emergencyContact?: { name: string; phone: string; relation?: string };
  specialNeeds?: string;
}

File: app/lib/models/participant.dart
class Participant {
  final String id;
  final String displayName;
  final String? firstName;
  final String? lastName;
  final int? age;
  final String? group;
  final int balance;
  final bool active;
  final String? nfcTagUid;
  final String eventId;

  const Participant({
    required this.id,
    required this.displayName,
    this.firstName,
    this.lastName,
    this.age,
    this.group,
    required this.balance,
    required this.active,
    this.nfcTagUid,
    required this.eventId,
  });

  factory Participant.fromJson(Map<String, dynamic> json) => Participant(
        id: json['id'] as String,
        displayName: json['displayName'] as String,
        firstName: json['firstName'] as String?,
        lastName: json['lastName'] as String?,
        age: json['age'] as int?,
        group: json['group'] as String?,
        balance: json['balance'] as int? ?? 0,
        active: json['active'] as bool? ?? true,
        nfcTagUid: (json['nfcTag'] as Map<String, dynamic>?)?['tagUid'] as String?,
        eventId: json['eventId'] as String? ?? json['event']?['id'] as String? ?? '',
      );

  String get fullName {
    final parts = [firstName, lastName].where((s) => s != null && s.isNotEmpty);
    return parts.isEmpty ? displayName : parts.join(' ');
  }
}
